stdlib/LinearAlgebra/src/dense.jl

```diff
@@ -92,6 +92,52 @@ isposdef(A::AbstractMatrix) =
     ishermitian(A) && isposdef(cholesky(Hermitian(A); check = false))
 isposdef(x::Number) = imag(x)==0 && real(x) > 0
 
+"""
+ispossemdef(A, k) -> Bool
+
+Test whether a matrix is positive semi-definite with specified rank `k` by
+checking that `k` of its eigenvalues are positive and the rest are zero.
+
+# Examples
+```jldoctest
+julia> A = [1 0; 0 0]
+2×2 Array{Int64,2}:
+ 1  0
+ 0  0
+
+julia> ispossemdef(A, 1)
+true
+
+julia> ispossemdef(A, 2)
+false
+"""
+function ispossemdef(X::AbstractMatrix, k::Int;
+                     atol::Real = 0.0,
+                     rtol::Real = (size(X, 1)*eps(real(float(one(eltype(X))))))*iszero(atol))
+    !ishermitian(X) && return false
+    !(0 <= k <= size(X, 1)) && error("rank must be in [0, n]")
```

Comment:
It may be hard to understand what `n` means in a longer program throwing this error, so maybe interpolate the value of `size(X, 1)` into the error message?
```julia
 !(0 <= k <= size(X, 1)) && error("rank must be between 0 and $(size(X, 1))")
```

Reply:
Done. Thanks!